File: src/element/structure.ts
import { MatchResult } from "../match";
import { Parser } from "../parser";
import { Pass } from "../pass/pass";
import { StructureStatement } from "../statement/structure";
import { Element, ElementPriority, ElementType } from "./element";

export abstract class Structure implements Element {
    public getPriority(): ElementPriority {
        return ElementPriority.NORMAL;
    }

    public getElementType(): ElementType {
        return ElementType.STRUCTURE;
    }

    public abstract visit(pass: Pass, statement: StructureStatement): boolean;
    public abstract initialize();
}

File: common/src/main/kotlin/me/honkling/skriptkt/common/syntax/Structure.kt
package me.honkling.skriptkt.common.syntax

import me.honkling.skriptkt.common.parser.Parser
import me.honkling.skriptkt.common.statement.Statement
import me.honkling.skriptkt.common.statement.StructureStatement

abstract class Structure : SyntaxElement() {
    abstract fun accept(statement: StructureStatement)
    abstract fun initialize(parser: Parser, statement: StructureStatement): Boolean

    override fun accept(statement: Statement) {
        if (statement !is StructureStatement)
            throw IllegalStateException("Passed non-structure statement to structure")

        accept(statement)
    }

    override fun initialize(parser: Parser, statement: Statement): Boolean {
        if (statement !is StructureStatement)
            throw IllegalStateException("Passed non-structure statement to structure")

        return initialize(parser, statement)
    }
}